File: src/main/java/com/nkming/powermenu/MainActivity.java
/*
 * MainActivity.java
 *
 * Author: Ming Tsang
 * Copyright (c) 2015 Ming Tsang
 * Refer to LICENSE for details
 */

package com.nkming.powermenu;

import android.app.KeyguardManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.ActionBarActivity;
import android.view.WindowManager;

public class MainActivity extends ActionBarActivity
{
	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		KeyguardManager km = (KeyguardManager)getSystemService(
				Context.KEYGUARD_SERVICE);
		if (km.inKeyguardRestrictedInputMode())
		{
			onLaunchWithKeyguard();
		}

		super.onCreate(savedInstanceState);
		PreferenceManager.setDefaultValues(this, getString(R.string.pref_file),
				Context.MODE_PRIVATE, R.xml.preference, false);

		if (!InstallHelper.isSystemApp(this))
		{
			// We don't want the animation here
			super.finish();
			startActivity(new Intent(this, InstallActivity.class));
		}
		else
		{
			// Disable the standard activity launch animation
			overridePendingTransition(0, 0);
			setContentView(R.layout.activity_main);
			if (savedInstanceState == null)
			{
				getSupportFragmentManager().beginTransaction()
						.add(R.id.container, new MainFragment()).commit();
			}
		}
	}

	@Override
	protected void onStart()
	{
		super.onStart();
		if (PersistentService.isRunning())
		{
			PersistentService.hideView(this);
		}
	}

	@Override
	protected void onStop()
	{
		super.onStop();
		if (PersistentService.isRunning())
		{
			PersistentService.showView(this);
		}
		else
		{
			PersistentViewHelper.startIfNecessary(this);
			SystemOverrideService.startIfNecessary(this);
		}
	}

	@Override
	protected void onUserLeaveHint()
	{
		super.onUserLeaveHint();
		if (mIsAnimateClose)
		{
			overridePendingTransition(0, R.anim.activity_close_exit);
		}
	}

	@Override
	public void finish()
	{
		super.finish();
		if (mIsAnimateClose)
		{
			overridePendingTransition(0, R.anim.activity_close_exit);
		}
	}

	private static final String LOG_TAG = MainActivity.class.getCanonicalName();

	/**
	 * Called when the activity is launched with an active keyguard
	 */
	private void onLaunchWithKeyguard()
	{
		Log.d(LOG_TAG, "onLaunchWithKeyguard");
		getWindow().addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED);
		setTheme(R.style.AppThemeKeyguard);
	}

	private boolean mIsAnimateClose = true;
}

File: src/main/kotlin/com/nkming/powermenu/MainActivity.kt
package com.nkming.powermenu

import android.app.KeyguardManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity()
{
	companion object
	{
		private val LOG_TAG = MainActivity::class.java.canonicalName
	}

	override fun onCreate(savedInstanceState: Bundle?)
	{
		_themeAdapter.onCreate(savedInstanceState)
		val km = getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
		_isKeyguard = km.inKeyguardRestrictedInputMode()
		if (_isKeyguard)
		{
			_onLaunchWithKeyguard()
		}

		super.onCreate(savedInstanceState)
		if (!InstallHelper.isSystemApp(this))
		{
			InstallHelper.isPowerCommandAvailable(this,
					onResult = {_onPostCheckInstall(savedInstanceState, it)})
		}
		else
		{
			_onPostCheckInstall(savedInstanceState, true)
		}
	}

	override fun onStart()
	{
		super.onStart()
		if (com.nkming.utils.widget.PersistentService.isRunning())
		{
			PersistentService.hideView(this)
		}
		_isStop = false
	}

	override fun onResume()
	{
		super.onResume()
		_themeAdapter.onResume()
	}

	override fun onPause()
	{
		super.onPause()
		_themeAdapter.onPause()
	}

	override fun onStop()
	{
		super.onStop()
		if (com.nkming.utils.widget.PersistentService.isRunning())
		{
			PersistentService.showView(this)
		}
		else
		{
			PersistentViewHelper.startIfNecessary(this)
			SystemOverrideService.startIfNecessary(this)
		}
		_isStop = true
	}

	override fun onDestroy()
	{
		super.onDestroy()
		_themeAdapter.onDestroy()
	}

	override fun onUserLeaveHint()
	{
		super.onUserLeaveHint()
		if (_isAnimateClose)
		{
			overridePendingTransition(0, R.anim.activity_close_exit)
		}
	}

	override fun finish()
	{
		super.finish()
		if (_isAnimateClose)
		{
			overridePendingTransition(0, R.anim.activity_close_exit)
		}
	}

	private fun _onPostCheckInstall(savedInstanceState: Bundle?,
			hasInstalled: Boolean)
	{
		if (!hasInstalled)
		{
			// We don't want the animation here
			super.finish()
			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
			{
				// System app not recommended for M+
				Toast.makeText(this, R.string.su_failed, Toast.LENGTH_LONG).show()
			}
			else
			{
				startActivity(Intent(this, InstallActivity::class.java))
			}
		}
		else
		{
			if (_isStop)
			{
				// User has closed/hid the app
				super.finish()
			}
			else
			{
				// Disable the standard activity launch animation
				overridePendingTransition(0, 0)
				setContentView(R.layout.activity_main)
				if (savedInstanceState == null)
				{
					supportFragmentManager.beginTransaction()
							.add(R.id.container, MainFragment())
							.commitAllowingStateLoss()
				}
			}
		}
	}

	/**
	 * Called when the activity is launched with an active keyguard
	 */
	private fun _onLaunchWithKeyguard()
	{
		Log.d(LOG_TAG, "_onLaunchWithKeyguard")
		window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
	}

	private var _isAnimateClose = true
	private var _isStop = false
	private var _isKeyguard = false

	private val _themeAdapter by lazy{ActivityThemeAdapter(this,
			R.style.AppTheme_Dark, R.style.AppTheme_Light)}
}
